sycl/source/builtins_generator.py

```diff
@@ -0,0 +1,1285 @@
+#   definition.
+class InstantiatedTemplatedArg:
+  def __init__(self, template_name, templated_type):
+    self.template_name = template_name
+    self.templated_type = templated_type
+
+  def __str__(self):
+    return self.template_name
+
+# A class representing an instantiated template return type, i.e. an association
+# between the return type of a builtin and a related template argument.
+# related_arg_type - The related template argument type.
+# template_type - The associated type picked from the valid types in the builtin
+#   definition.
+class InstantiatedTemplatedReturnType:
+  def __init__(self, related_arg_type, templated_type):
+    self.related_arg_type = related_arg_type
+    self.templated_type = templated_type
+
+  def __str__(self):
+    if isinstance(self.templated_type, Vec):
+      # Vectors need conversion as their related type may be a swizzle.
+      return f'detail::get_vec_t<{self.related_arg_type}>'
+    return str(self.related_arg_type)
+
+# A class representing an instantiated version of ElementType containing the
+# referenced type.
+# referenced_type - The referenced type.
+# parent_idx - The index of the parent argument in the corresponding builtin.
+class InstantiatedElementType:
+  def __init__(self, referenced_type, parent_idx):
+    self.referenced_type = referenced_type
+    self.parent_idx = parent_idx
+
+  def __str__(self):
+    if (isinstance(self.referenced_type, InstantiatedTemplatedArg) or
+        isinstance(self.referenced_type, InstantiatedTemplatedReturnType)):
+      return f'detail::get_elem_type_t<{self.referenced_type}>'
+    return self.referenced_type
+
+# A class representing an instantiated version of ConversionTraitType containing
+# the parent type.
+# parent_type - The parent type.
+# trait - The name of the trait to use for converting the type.
+# parent_idx - The index of the parent argument in the corresponding builtin.
+class InstantiatedConversionTraitType:
+  def __init__(self, parent_type, trait, parent_idx):
+    self.parent_type = parent_type
+    self.trait = trait
+    self.parent_idx = parent_idx
+
+  def __str__(self):
+    return f'detail::{self.trait}<{self.parent_type}>'
+
+### GENTYPE DEFINITIONS
+
+vfloatn = [Vec(["float"])]
+vfloat3or4 = [Vec(["float"], {3,4})]
+mfloatn = [Marray(["float"])]
+mfloat3or4 = [Marray(["float"], {3,4})]
+genfloatf = ["float", Vec(["float"]), Marray(["float"])]
+
+vdoublen = [Vec(["double"])]
+vdouble3or4 = [Vec(["double"], {3,4})]
+mdoublen = [Marray(["double"])]
+mdouble3or4 = [Marray(["double"], {3,4})]
+
+vhalfn = [Vec(["half"])]
+vhalf3or4 = [Vec(["half"], {3,4})]
+mhalfn = [Marray(["half"])]
+mhalf3or4 = [Marray(["half"], {3,4})]
+
+genfloat = ["float", "double", "half", Vec(["float", "double", "half"]),
+            Marray(["float", "double", "half"])]
+vgenfloat = [Vec(["float", "double", "half"])]
+sgenfloat = ["float", "double", "half"]
+mgenfloat = [Marray(["float", "double", "half"])]
```

Comment:
```suggestion
sgenfloat = ["float", "double", "half"]
vgenfloat = [Vec(sgenfloat)]
mgenfloat = [Marray(sgenfloat)]
genfloat = sgenfloat + vgenfloat + mgenfloat
```

Lists can be [concatenated](https://docs.python.org/3/library/stdtypes.html#common-sequence-operations)

Reply:
I intentionally avoided concatenating the groups, primarily to make them 1:1 with how the specification defines them and to make it easier for people to look up which types a given definition should map to.

Reply:
I don't quite understand how concatenation would break `1:1` mapping: to me, it is just a way to reduce verbosity by re-using existing definitions and built upon them.

From [Table 166](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#table.gentypes):

> *genfloat*
> genfloatf, genfloatd, genfloath
>
> *genfloatf*
> float, floatn
>
> *floatn*
> float{n}, mfloat{n}, marray<{N},float>

Two observations:
1. The code snippet above is already not 1:1 mapping, because it does not define `genfloatf`;
2. Spec does re-use previously defined pseudo-types in the example above (even though I agree, not in a way I'm suggesting)